DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/register_map_and_streaming_interfaces/README.md

```diff
@@ -0,0 +1,324 @@
+  q.single_task<LambdaRegisterMapControlIP>([=] register_map_interface  {
+    for (int i = 0; i < count; i++) {
+      out[i] = SomethingComplicated(in[i]);
+    }
+  }).wait();
+}
+
+void TestLambdaStreamingControlKernel(sycl::queue &q, ValueT *in, ValueT *out, size_t count) {
+  q.single_task<LambdaStreamingControlIP>([=] streaming_interface  {
+    for (int i = 0; i < count; i++) {
+      out[i] = SomethingComplicated(in[i]);
+    }
+  }).wait();
+}
+```
+
+## Key Concepts
+* Basics of declaring kernel control interfaces and kernel argument interfaces
+
+## Building the `register_map_and_streaming_interfaces` Tutorial
+
+> **Note**: If you have not already done so, set up your CLI
+> environment by sourcing  the `setvars` script located in
+> the root of your oneAPI installation.
+>
+> Linux Sudo: `. /opt/intel/oneapi/setvars.sh`
+>
+> Linux User: `. ~/intel/oneapi/setvars.sh`
+>
+> Windows: `C:\Program Files(x86)\Intel\oneAPI\setvars.bat`
+>
+>For more information on environment variables, see Use the setvars Script for [Linux or macOS](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html), or [Windows](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).
+
+### Using Visual Studio Code*  (Optional)
+
+You can use Visual Studio Code (VS Code) extensions to set your environment, create launch configurations,
+and browse and download samples.
+
+The basic steps to build and run a sample using VS Code include:
+ - Download a sample using the extension **Code Sample Browser for Intel oneAPI Toolkits**.
+ - Configure the oneAPI environment with the extension **Environment Configurator for Intel oneAPI Toolkits**.
+ - Open a Terminal in VS Code (**Terminal>New Terminal**).
+ - Run the sample in the VS Code terminal using the instructions below.
+
+To learn more about the extensions and how to configure the oneAPI environment, see
+[Using Visual Studio Code with Intel® oneAPI Toolkits](https://software.intel.com/content/www/us/en/develop/documentation/using-vs-code-with-intel-oneapi/top.html).
+
+After learning how to use the extensions for Intel oneAPI Toolkits, return to this readme for instructions on how to build and run a sample.
+
+### On a Linux* System
+
+1. Generate the Makefile by running `cmake`.
+     ```
+   mkdir build
+   cd build
+   ```
+   To compile for the Intel® Arria10® FPGA family, run `cmake` using the following command:
+
+   ```
+   cmake ..
+   ```
+   
+   You can also compile for a custom FPGA platform. Ensure that the board support package is installed on your system. Then run `cmake` using the following command:
+   
+   ```
+   cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
```

Comment:
This is not really true for IPA
FPGA_DEVICE should either be an FPGA family or an FPGA part right?

Reply:
Please also split these kernels up so the user can see the interfaces in the generated RTL. The way it's written now, they all get munged together and it's hard to understand what's going on. Options:

1. use `-fsycl-device-code-split=per_kernel`
1. multiple C++ files, each generating a separate kernel

I'm personally in favour of the second one because `-fsycl-device-code-split=per_kernel` is the cause of much frustration. Option 2 will also make adding questa waveforms as I've mentioned below much easier.

Reply:
Updated